File: backend/convex/Memory/extractEmotion.ts
const { AutoTokenizer, AutoModelForSequenceClassification } = require('@xenova/transformers');

// Load the tokenizer and model
async function loadModel() {
    const tokenizer = await AutoTokenizer.from_pretrained('michellejieli/emotion_text_classifier');
    const model = await AutoModelForSequenceClassification.from_pretrained('michellejieli/emotion_text_classifier');
    return { tokenizer, model };
}

// Function to predict emotion from text
async function predictEmotion(text, tokenizer, model) {
    // Tokenize input text
    const inputs = await tokenizer(text, { return_tensors: 'pt' });

    // Get the predictions
    const outputs = await model(inputs);
    const logits = outputs.logits;

    // Apply softmax to get probabilities
    const softmax = (x) => {
        const maxLogit = Math.max(...x);
        const exps = x.map((i) => Math.exp(i - maxLogit));
        const sumExps = exps.reduce((a, b) => a + b);
        return exps.map((i) => i / sumExps);
    };

    const probabilities = softmax(logits.data[0]);

    // Get the predicted emotion
    const predictedClassId = probabilities.indexOf(Math.max(...probabilities));
    const predictedEmotion = model.config.id2label[predictedClassId];

    // Convert to dictionary
    const outputDict = {
        text: text,
        predicted_emotion: predictedEmotion,
        probabilities: model.config.id2label.reduce((obj, label, idx) => {
            obj[label] = probabilities[idx];
            return obj;
        }, {})
    };

    return outputDict;
}

// Example usage
(async () => {
    const { tokenizer, model } = await loadModel();
    const text = 'I am so happy today!';
    const result = await predictEmotion(text, tokenizer, model);
    console.log(result);
})();

File: backend/PerceptionAnalysis/extract_emotion.py
from transformers import AutoTokenizer, AutoModelForSequenceClassification
import torch
import torch.nn.functional as F

# Load the tokenizer and model
tokenizer = AutoTokenizer.from_pretrained("michellejieli/emotion_text_classifier")
model = AutoModelForSequenceClassification.from_pretrained("michellejieli/emotion_text_classifier")

# Function to predict emotion from text
def predict_emotion(text):
    # Tokenize input text
    inputs = tokenizer(text, return_tensors='pt')
    # Get the predictions
    with torch.no_grad():
        outputs = model(**inputs)
    logits = outputs.logits
    # Apply softmax to get probabilities
    probabilities = F.softmax(logits, dim=-1)
    # Get the predicted emotion
    predicted_class_id = probabilities.argmax(axis=-1).item()
    predicted_emotion = model.config.id2label[predicted_class_id]
    # Convert to dictionary
    output_dict = {
        "text": text,
        "predicted_emotion": predicted_emotion,
        "probabilities": {model.config.id2label[i]: prob.item() for i, prob in enumerate(probabilities[0])}
    }
    return output_dict

# Example usage
text = ("Yesterday I worked online most of the day. I discovered an assignment I had done "
        "the previous day was rejected. This assignment provided a considerable reward so I "
        "basically lost about an hour's worth of work. That was frustrating. We need the money "
        "I am making online right now to pay our bills so I was also concerned about making enough "
        "money. However, at the end of the day, I hit my earnings goal so I was satisfied with my progress.")


if __name__ == "__main__":
    import sys
    text = sys.argv[1] if len(sys.argv) > 1 else ""
    if text:
        print(predict_emotion(text))
    else:
        print("No text provided")
